# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/cache/DBSchema.cpp
# Commit: 2fb182664097
# Full Hash: 2fb182664097f1b22021eec844d0d1adbffc40dc
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2020-12-05 09:38:58
# Regressor Bug: 1678030
# File Overlap Count: 1
# Description:
#   Bug 1678030 - Support multiple variants of hasResult expectations in CreateAndExecuteSingleStepStatement. r=dom-workers-and-storage-reviewers,ttung
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D97461
# ==============================================================================

diff -r 2ed5e139d784 -r 2fb182664097 dom/cache/DBSchema.cpp
--- a/dom/cache/DBSchema.cpp	Fri Dec 04 14:56:18 2020 +0000
+++ b/dom/cache/DBSchema.cpp	Fri Dec 04 15:10:07 2020 +0000
@@ -645,21 +645,12 @@
     return Err(rv);
   }
 
-  nsCOMPtr<mozIStorageStatement> state;
-  rv = aConn.CreateStatement("SELECT last_insert_rowid()"_ns,
-                             getter_AddRefs(state));
-  if (NS_WARN_IF(NS_FAILED(rv))) {
-    return Err(rv);
-  }
-
-  bool hasMoreData = false;
-  rv = state->ExecuteStep(&hasMoreData);
-  if (NS_WARN_IF(NS_FAILED(rv))) {
-    return Err(rv);
-  }
-  if (NS_WARN_IF(!hasMoreData)) {
-    return Err(NS_ERROR_UNEXPECTED);
-  }
+  CACHE_TRY_INSPECT(const auto& state,
+                    quota::CreateAndExecuteSingleStepStatement<
+                        quota::SingleStepResult::ReturnNullIfNoResult>(
+                        aConn, "SELECT last_insert_rowid()"_ns));
+
+  CACHE_TRY(OkIf(state), Err(NS_ERROR_UNEXPECTED));
 
   CacheId id;
   rv = state->GetInt64(0, &id);
@@ -2066,7 +2057,7 @@
                           quota::CreateAndExecuteSingleStepStatement(
                               aConn, "SELECT last_insert_rowid()"_ns));
 
-        CACHE_TRY_RETURN(MOZ_TO_RESULT_INVOKE(state, GetInt32, 0));
+        CACHE_TRY_RETURN(MOZ_TO_RESULT_INVOKE(*state, GetInt32, 0));
       }()));
 
   rv = aConn.CreateStatement(
@@ -2727,21 +2718,14 @@
     // (pragma_table_info is a table-valued function format variant of
     // "PRAGMA table_info" supported since SQLite 3.16.0.  Firefox 53 shipped
     // was the first release with this functionality, shipping 3.16.2.)
-    nsCOMPtr<mozIStorageStatement> stmt;
-    rv = aConn.CreateStatement(
-        nsLiteralCString("SELECT name FROM pragma_table_info('entries') WHERE "
-                         "name = 'response_padding_size'"),
-        getter_AddRefs(stmt));
-    if (NS_WARN_IF(NS_FAILED(rv))) {
-      return Err(rv);
-    }
-
+    //
     // If there are any result rows, then the column is present.
-    bool hasColumn = false;
-    rv = stmt->ExecuteStep(&hasColumn);
-    if (NS_WARN_IF(NS_FAILED(rv))) {
-      return Err(rv);
-    }
+    CACHE_TRY_INSPECT(const bool& hasColumn,
+                      quota::CreateAndExecuteSingleStepStatement<
+                          quota::SingleStepResult::ReturnNullIfNoResult>(
+                          aConn,
+                          "SELECT name FROM pragma_table_info('entries') WHERE "
+                          "name = 'response_padding_size'"_ns));
 
     if (hasColumn) {
       return kHackyPaddingSizePresentVersion;
@@ -3116,21 +3100,12 @@
 
   // Revalidate the foreign key constraints, and ensure that there are no
   // violations.
-  nsCOMPtr<mozIStorageStatement> state;
-  rv = aConn.CreateStatement("PRAGMA foreign_key_check;"_ns,
-                             getter_AddRefs(state));
-  if (NS_WARN_IF(NS_FAILED(rv))) {
-    return rv;
-  }
-
-  bool hasMoreData = false;
-  rv = state->ExecuteStep(&hasMoreData);
-  if (NS_WARN_IF(NS_FAILED(rv))) {
-    return rv;
-  }
-  if (NS_WARN_IF(hasMoreData)) {
-    return NS_ERROR_FAILURE;
-  }
+  CACHE_TRY_INSPECT(const bool& hasResult,
+                    quota::CreateAndExecuteSingleStepStatement<
+                        quota::SingleStepResult::ReturnNullIfNoResult>(
+                        aConn, "PRAGMA foreign_key_check;"_ns));
+
+  CACHE_TRY(OkIf(!hasResult), NS_ERROR_FAILURE);
 
   rv = aConn.SetSchemaVersion(17);
   if (NS_WARN_IF(NS_FAILED(rv))) {
@@ -3360,21 +3335,12 @@
 
   // Revalidate the foreign key constraints, and ensure that there are no
   // violations.
-  nsCOMPtr<mozIStorageStatement> state;
-  rv = aConn.CreateStatement("PRAGMA foreign_key_check;"_ns,
-                             getter_AddRefs(state));
-  if (NS_WARN_IF(NS_FAILED(rv))) {
-    return rv;
-  }
-
-  bool hasMoreData = false;
-  rv = state->ExecuteStep(&hasMoreData);
-  if (NS_WARN_IF(NS_FAILED(rv))) {
-    return rv;
-  }
-  if (NS_WARN_IF(hasMoreData)) {
-    return NS_ERROR_FAILURE;
-  }
+  CACHE_TRY_INSPECT(const bool& hasResult,
+                    quota::CreateAndExecuteSingleStepStatement<
+                        quota::SingleStepResult::ReturnNullIfNoResult>(
+                        aConn, "PRAGMA foreign_key_check;"_ns));
+
+  CACHE_TRY(OkIf(!hasResult), NS_ERROR_FAILURE);
 
   rv = aConn.SetSchemaVersion(21);
   if (NS_WARN_IF(NS_FAILED(rv))) {